modules/sdk/src/main/scala/org/tessellation/sdk/http/p2p/middlewares/PeerAuthMiddleware.scala

```diff
@@ -0,0 +1,217 @@
+    privateKey: PrivateKey,
+    sessionStorage: SessionStorage[F],
+    selfId: PeerId
+  ): Client[F] = Client { req: Request[F] =>
+    val signer = new Http4sRequestSigner(getGenerator(privateKey), new TessellationHttpCryptoConfig {})
+
+    Resource.suspend {
+      Ref[F].of(Vector.empty[Chunk[Byte]]).map { vec =>
+        Resource.liftK[F] {
+
+          val copiedBody = Stream
+            .eval(vec.get)
+            .flatMap(v => Stream.emits(v).covary[F])
+            .flatMap(c => Stream.chunk(c).covary[F])
+
+          for {
+            tokenHeader <- sessionStorage.getToken.map(_.map(t => `X-Session-Token`(t)))
+            newHeaders = tokenHeader.fold(req.headers)(h => req.headers.put(h)).put(`X-Id`(selfId))
+            newReq = req
+              .withBodyStream(req.body.observe(_.chunks.flatMap(s => Stream.exec(vec.update(_ :+ s)))))
+              .withHeaders(newHeaders)
+            signedRequest <- signer.sign(newReq).map(r => req.withBodyStream(copiedBody).withHeaders(r.headers))
+          } yield signedRequest
+
+        }
+      }
+    } >>= client.run
+  }
+
+  def requestTokenVerifierMiddleware[F[_]: Async](session: Session[F])(http: HttpRoutes[F]): HttpRoutes[F] =
+    Kleisli { req: Request[F] =>
+      val host = req.uri.authority.map(_.host.value).flatMap(Host.fromString)
+      val port = req.uri.authority.flatMap(_.port).flatMap(Port.fromInt)
+      val token = req.headers
+        .get[`X-Session-Token`]
+        .map(_.token)
+
+      Applicative[Option]
+        .map2(host, port) { (_host, _port) =>
+          session.verifyToken(_host, _port, token).attemptT.toOption.flatMap {
+            case TokenValid => http(req)
+            case _          => OptionT.pure[F](unauthorized[F])
+          }
+        }
+        .getOrElse(OptionT.pure[F](unauthorized[F]))
+    }
+
+  def requestVerifierMiddleware[F[_]: Async: SecurityProvider](
+    knownPeer: (Host, Port) => F[Option[PeerId]]
+  )(http: HttpRoutes[F]): HttpRoutes[F] =
+    Kleisli { req: Request[F] =>
+      val host = req.uri.authority.map(_.host.value).flatMap(Host.fromString)
+      val port = req.uri.authority.flatMap(_.port).flatMap(Port.fromInt)
+      val idFromHeaders = req.headers.get[`X-Id`].map(_.id)
+
+      val verify: OptionT[F, Response[F]] = for {
+
+        tuple <- Ref[F]
+          .of(Vector.empty[Chunk[Byte]])
+          .map { vec =>
+            val newBody =
+              Stream.eval(vec.get).flatMap(v => Stream.emits(v).covary[F]).flatMap(c => Stream.chunk(c).covary[F])
+            val newReq = req.withBodyStream(req.body.observe(_.chunks.flatMap(s => Stream.exec(vec.update(_ :+ s)))))
+            (newBody, newReq)
+          }
+          .attemptT
+          .toOption
+
+        id <- Applicative[Option]
+          .map2(host, port) { (_host, _port) =>
+            knownPeer(_host, _port).map(_.orElse(idFromHeaders)).attemptT.toOption.flatMap(_.toOptionT)
+          }
+          .getOrElse(OptionT.none)
```

Comment:
Is it then `Option[OptionT[Id]]`? Can we just `flatMap` it instead of `getOrElse`?

Reply:
it's `Option` and `OptionT` and `OptionT` is `F[Option[..]]` so it's `Option[F[Option` at the end I think
I will take a look if I can traverse and flatMap